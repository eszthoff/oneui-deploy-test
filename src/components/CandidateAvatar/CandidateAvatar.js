import React from 'react';
import PropTypes from 'prop-types';
import bem from '../../packages/bem';
import styles from './CandidateAvatar.scss';

const { block, elem } = bem({
    name: 'CandidateAvatar',
    classnames: styles,
    propsToMods: ['context']
});

const context = percentage => {
    if (percentage <= 33) {
        return 'bad';
    }
    if (percentage <= 66) {
        return 'warning';
    }
    return 'good';
};

const CandidateAvatar = props => {
    const { imageUrl, matchPercentage, showPercentageOnHover, size, ...rest } = props;

    const fixedSize = Math.round(Math.max(32, size));
    const radius = (fixedSize / 2) - 2;
    const percentage = Math.max(0, Math.min(100, matchPercentage));
    const circumference = 2 * radius * Math.PI;
    const strokeDasharray = `${(percentage * circumference) / 100} 999`;

    return (
        <div {...rest} {...block(props)}
            style={{
                width: fixedSize,
                height: fixedSize
            }}
        >
            <div
                {...elem('image', props)}
                style={{
                    backgroundImage: `url(${imageUrl})`
                }}
            >
                { !!showPercentageOnHover && (
                    <div {...elem('percentage', props)}>{`${percentage}%`}</div>
                )}
            </div>
            <svg {...elem('ring', props)}
                style={{
                    width: fixedSize,
                    height: fixedSize
                }}
            >
                <circle
                    r={ radius}
                    cx={ fixedSize / 2}
                    cy={ fixedSize / 2 }
                    strokeWidth="4"
                    {...elem('circle', {
                        ...props,
                        context: context(percentage)
                    })}
                    style={{
                        strokeDasharray
                    }}
                />
            </svg>
        </div>
    );
};

CandidateAvatar.propTypes = {
    imageUrl: PropTypes.string,
    matchPercentage: PropTypes.number,
    showPercentageOnHover: PropTypes.bool,
    size: PropTypes.number
};

CandidateAvatar.defaultProps = {
    imageUrl: null,
    matchPercentage: null,
    showPercentageOnHover: false,
    size: 72
};

export default CandidateAvatar;
