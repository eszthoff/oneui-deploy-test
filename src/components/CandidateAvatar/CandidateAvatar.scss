.CandidateAvatar {
    position: relative;
    display: inline-block;
    height: 72px;
    width: 72px;

    &__ring {
        width: 72px;
        height: 72px;
    }

    &__circle {
        fill: none;
        stroke: var(--color-neutral);
        stroke-dasharray: 0 999;
        stroke-linecap: round;
        shape-rendering: geometricPrecision;
        transition: all .3s ease;
        transform: rotate(-90deg);
        transform-origin: 50% 50%;

        &--context {
            @each $context in (good, warning, bad) {
                &_#{$context} {
                    stroke: var(--color-#{$context});
                }
            }
        }
    }

    &__image {
        border-radius: 50%;
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        height: 60px;
        width: 60px;
        background-size: cover;
        background-position: 50% 50%;

        &:hover {
            .CandidateAvatar__percentage {
                opacity: 1;
            }
        }
    }

    &__percentage {
        background-color: rgba(255,255,255,.25);
        border-radius: 50%;
        display: flex;
        align-items: center;
        justify-content: center;
        height: 100%;
        width: 100%;
        opacity: 0;
        font: {
            family: var(--font-family-primary);
            size: var(--font-size-large);
        }
        transition: all .3s;
    }
}
