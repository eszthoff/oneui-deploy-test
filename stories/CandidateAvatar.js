import React from 'react';
import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import { boolean, number, text, withKnobs } from '@storybook/addon-knobs'; // eslint-disable-line import/no-extraneous-dependencies
import CandidateAvatar from '../src/components/CandidateAvatar';

storiesOf('CandidateAvatar', module)
    .addDecorator(withKnobs)
    .add('CandidateAvatar', () => (
        <CandidateAvatar
            imageUrl={text('Image URL', 'https://placehold.it/350x150')}
            matchPercentage={number('Match percentage', 75)}
            showPercentageOnHover={boolean('Show percentage on hover', true)}
            size={number('Size', 72)}
        />
    ));
