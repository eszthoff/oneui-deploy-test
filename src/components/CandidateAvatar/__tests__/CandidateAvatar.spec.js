import React from 'react';
import toJson from 'enzyme-to-json';
import CandidateAvatar from '../CandidateAvatar';

describe('<CandidateAvatar> that renders a candidate profile image with match indication', () => {
    it('should render a default avatar', () => {
        const wrapper = shallow(<CandidateAvatar />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render a custom sized avatar with good match percentage', () => {
        const wrapper = shallow(<CandidateAvatar size={128} percentage={67} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render an avatar with average match percentage', () => {
        const wrapper = shallow(<CandidateAvatar size={128} percentage={34} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render an avatar with bad match percentage', () => {
        const wrapper = shallow(<CandidateAvatar size={128} percentage={33} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should add classes when props are changed', () => {
        const wrapper = shallow(<CandidateAvatar showPercentageOnHover percentage={25} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
