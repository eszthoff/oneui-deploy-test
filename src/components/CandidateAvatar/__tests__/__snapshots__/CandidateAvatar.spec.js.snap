// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CandidateAvatar> that renders a candidate profile image with match indication should add classes when props are changed 1`] = `
<div
  className="CandidateAvatar"
  percentage={25}
  style={
    Object {
      "height": 72,
      "width": 72,
    }
  }
>
  <div
    className="CandidateAvatar__image"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  >
    <div
      className="CandidateAvatar__percentage"
    >
      0%
    </div>
  </div>
  <svg
    className="CandidateAvatar__ring"
    style={
      Object {
        "height": 72,
        "width": 72,
      }
    }
  >
    <circle
      className="CandidateAvatar__circle CandidateAvatar__circle--context_bad"
      cx={36}
      cy={36}
      r={34}
      strokeWidth="4"
      style={
        Object {
          "strokeDasharray": "0 999",
        }
      }
    />
  </svg>
</div>
`;

exports[`<CandidateAvatar> that renders a candidate profile image with match indication should render a custom sized avatar with good match percentage 1`] = `
<div
  className="CandidateAvatar"
  percentage={67}
  style={
    Object {
      "height": 128,
      "width": 128,
    }
  }
>
  <div
    className="CandidateAvatar__image"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  />
  <svg
    className="CandidateAvatar__ring"
    style={
      Object {
        "height": 128,
        "width": 128,
      }
    }
  >
    <circle
      className="CandidateAvatar__circle CandidateAvatar__circle--context_bad"
      cx={64}
      cy={64}
      r={62}
      strokeWidth="4"
      style={
        Object {
          "strokeDasharray": "0 999",
        }
      }
    />
  </svg>
</div>
`;

exports[`<CandidateAvatar> that renders a candidate profile image with match indication should render a default avatar 1`] = `
<div
  className="CandidateAvatar"
  style={
    Object {
      "height": 72,
      "width": 72,
    }
  }
>
  <div
    className="CandidateAvatar__image"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  />
  <svg
    className="CandidateAvatar__ring"
    style={
      Object {
        "height": 72,
        "width": 72,
      }
    }
  >
    <circle
      className="CandidateAvatar__circle CandidateAvatar__circle--context_bad"
      cx={36}
      cy={36}
      r={34}
      strokeWidth="4"
      style={
        Object {
          "strokeDasharray": "0 999",
        }
      }
    />
  </svg>
</div>
`;

exports[`<CandidateAvatar> that renders a candidate profile image with match indication should render an avatar with average match percentage 1`] = `
<div
  className="CandidateAvatar"
  percentage={34}
  style={
    Object {
      "height": 128,
      "width": 128,
    }
  }
>
  <div
    className="CandidateAvatar__image"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  />
  <svg
    className="CandidateAvatar__ring"
    style={
      Object {
        "height": 128,
        "width": 128,
      }
    }
  >
    <circle
      className="CandidateAvatar__circle CandidateAvatar__circle--context_bad"
      cx={64}
      cy={64}
      r={62}
      strokeWidth="4"
      style={
        Object {
          "strokeDasharray": "0 999",
        }
      }
    />
  </svg>
</div>
`;

exports[`<CandidateAvatar> that renders a candidate profile image with match indication should render an avatar with bad match percentage 1`] = `
<div
  className="CandidateAvatar"
  percentage={33}
  style={
    Object {
      "height": 128,
      "width": 128,
    }
  }
>
  <div
    className="CandidateAvatar__image"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  />
  <svg
    className="CandidateAvatar__ring"
    style={
      Object {
        "height": 128,
        "width": 128,
      }
    }
  >
    <circle
      className="CandidateAvatar__circle CandidateAvatar__circle--context_bad"
      cx={64}
      cy={64}
      r={62}
      strokeWidth="4"
      style={
        Object {
          "strokeDasharray": "0 999",
        }
      }
    />
  </svg>
</div>
`;
